src/loongarch64/ffi.c

```diff
@@ -0,0 +1,595 @@
+{
+  size_t realign[2];
+  void *pointer;
+
+#if ABI_FRLEN
+  if (!var && type->type == FFI_TYPE_STRUCT)
+    {
+      float_struct_info fsi = struct_passed_as_elements (cb, type);
+      if (fsi.as_elements)
+	{
+	  unmarshal_atom (cb, fsi.type1, data);
+	  if (fsi.offset2)
+	    unmarshal_atom (cb, fsi.type2, ((char *) data) + fsi.offset2);
+	  return data;
+	}
+    }
+
+  if (!var && cb->used_float < NARGREG
+      && IS_FLOAT (type->type))
+    {
+      unmarshal_atom (cb, type->type, data);
+      return data;
+    }
+
+  if (var && type->type == FFI_TYPE_FLOAT)
+    {
+      int m = cb->used_integer;
+      void *promoted
+	= m < NARGREG ? cb->aregs->a + m : cb->used_stack + m - NARGREG + 1;
+      *(float *) promoted = *(double *) promoted;
+    }
+#endif
+
+  if (type->size > 2 * __SIZEOF_POINTER__)
+    {
+      /* Pass by reference.  */
+      unmarshal_atom (cb, FFI_TYPE_POINTER, (char *) &pointer);
+      return pointer;
+    }
+  else if (IS_INT (type->type) || type->type == FFI_TYPE_POINTER)
+    {
+      unmarshal_atom (cb, type->type, data);
+      return data;
+    }
+  else
+    {
+      /* Overlong integers, soft-float floats, and structs without special
+	 float handling are treated identically from this point on.  */
+
+      /* Variadics are aligned even in registers.  */
+      if (type->alignment > __SIZEOF_POINTER__)
+	{
+	  if (var)
+	    cb->used_integer = FFI_ALIGN (cb->used_integer, 2);
+	  cb->used_stack
+	    = (size_t *) FFI_ALIGN (cb->used_stack, 2 * __SIZEOF_POINTER__);
+	}
+
+      if (type->size > 0)
+	unmarshal_atom (cb, FFI_TYPE_POINTER, realign);
+      if (type->size > __SIZEOF_POINTER__)
+	unmarshal_atom (cb, FFI_TYPE_POINTER, realign + 1);
+      memcpy (data, realign, type->size);
+      return data;
+    }
+}
+
+static int
+passed_by_ref (call_builder *cb, ffi_type *type, int var)
+{
+#if ABI_FRLEN
+  if (!var && type->type == FFI_TYPE_STRUCT)
+    {
+      float_struct_info fsi = struct_passed_as_elements (cb, type);
+      if (fsi.as_elements)
+	return 0;
+    }
+#endif
+
+  return type->size > 2 * __SIZEOF_POINTER__;
+}
+
+/* Perform machine dependent cif processing.  */
+ffi_status
+ffi_prep_cif_machdep (ffi_cif *cif)
+{
+  cif->loongarch_nfixedargs = cif->nargs;
+  return FFI_OK;
+}
+
+/* Perform machine dependent cif processing when we have a variadic
+   function.  */
+ffi_status
+ffi_prep_cif_machdep_var (ffi_cif *cif, unsigned int nfixedargs,
+			  unsigned int ntotalargs)
+{
+  cif->loongarch_nfixedargs = nfixedargs;
+  return FFI_OK;
+}
+
+/* Low level routine for calling functions.  */
+extern void ffi_call_asm (void *stack, struct call_context *regs,
+			  void (*fn) (void), void *closure) FFI_HIDDEN;
+
+static void
+ffi_call_int (ffi_cif *cif, void (*fn) (void), void *rvalue, void **avalue,
+	      void *closure)
+{
+  /* This is a conservative estimate, assuming a complex return value and
+     that all remaining arguments are long long / __int128 */
+  size_t arg_bytes = cif->bytes;
```

Comment:
> Where is this `bytes` initialized? I see in the original arm64 code it's got an assignment somewhere else, while here you seem to have dropped the original logic (from RISC-V) without replacement. Also the comment above is no longer accurate.

I don't know how to change the code correctly, can you give me some hints. Thanks

Reply:
> > Where is this `bytes` initialized? I see in the original arm64 code it's got an assignment somewhere else, while here you seem to have dropped the original logic (from RISC-V) without replacement. Also the comment above is no longer accurate.
> 
> I don't know how to change the code correctly, can you give me some hints. Thanks

I'll have to reply later, maybe tonight or tomorrow, because I also haven't looked at this more closely, only a cursory look yesterday, and also because I'm doing all LoongArch work in my free time. In the meantime maybe other people can help you too.

Reply:
> In the meantime maybe other people can help you too.

Thanks for your help.

Reply:
Initialized here: https://github.com/libffi/libffi/blob/9af445fd70e4d7f2b3bdfc511cc3cc76a6b84c9a/src/prep_cif.c#L210

But I'm not sure if we need to modify it for a 16-byte stack alignment like aarch64 in `ffi_prep_cif_machdep`.

Reply:
> Initialized here:
> 
> https://github.com/libffi/libffi/blob/9af445fd70e4d7f2b3bdfc511cc3cc76a6b84c9a/src/prep_cif.c#L210
> 
> But I'm not sure if we need to modify it for a 16-byte stack alignment like aarch64 in `ffi_prep_cif_machdep`.

I hope libffi will be merged into the upstream as soon as possible，there are many software dependencies on libffi on Loongarch